### DISCUSSION TITLE
How to setQueryData without exact key

### DISCUSSION THREAD

**Discussion by YassienW** [TanStack/query#2078] [Q&A] [answered]
According to https://github.com/tannerlinsley/react-query/discussions/394, query key was changed to be exact by default. Is there a way to not match a query key exactly when using setQueryData?

**Comment by TkDodo** [comment]
no, when you call `setQueryData`, the value that you pass in will be set for the given key. I'm not sure how that should work fuzzily, as it doesn't care about existing entries in the cache

**Comment by TkDodo** [comment]
I just had this use case as well when updating a list query where I have multiple cache entries due to filters, and this is what I’ve come up with:

```
queryClient
  .getQueryCache()
  .findAll(['partial', 'key'])
  .forEach(({ queryKey }) => {
    queryClient.setQueryData(queryKey, newData)
  })
```

**Reply by YassienW** [reply]
Thanks for sharing. I ended up getting rid of the key array, but i'm sure this will come up again since having filters in a query key is pretty common. I do think there should be support for this in the api

**Reply by TkDodo** [reply]
I might add it since it’s just 3 lines :)

**Reply by villesau** [reply, marked as answer]
There seems to be now even easier way: [setQueriesData](https://react-query.tanstack.com/reference/QueryClient#queryclientsetqueriesdata)

**Reply by Rc85** [reply]
With setQueriesData, setQueryData is redundant.

**Reply by bartvanandel** [reply]
> There seems to be now even easier way: [setQueriesData](https://react-query.tanstack.com/reference/QueryClient#queryclientsetqueriesdata)

Unfortunately, the updater function in `setQueriesData` doesn't currently have access to the exact query key. This makes it hard to update e.g. filtered queries where the filter is part of the query key.

Illustrated (pseudo code):
```ts
// Somewhere in a component
const filteredItems = useQuery({ queryKey: ['itemsByState', { state: ['state1', 'state2'] }], ... })
const moreFilteredItems = useQuery({ queryKey: ['itemsByState', { state: ['state1', 'state3'] }], ... })

// Potentially somewhere else
const updatedItem = { ...filteredItems[0], state: 'state3' }

queryClient.setQueriesData({ queryKey: ['itemsByState'], exact: false }, (prev) => {
  // Which filtered items are we updating here? Can't tell!
})
```

Current function and type definitions:

```ts
setQueriesData<TQueryFnData>(queryKey: QueryKey, updater: Updater<TQueryFnData | undefined, TQueryFnData | undefined>, options?: SetDataOptions): [QueryKey, TQueryFnData | undefined][];
setQueriesData<TQueryFnData>(filters: QueryFilters, updater: Updater<TQueryFnData | undefined, TQueryFnData | undefined>, options?: SetDataOptions): [QueryKey, TQueryFnData | undefined][];

export declare type Updater<TInput, TOutput> = TOutput | DataUpdateFunction<TInput, TOutput>;
export declare type DataUpdateFunction<TInput, TOutput> = (input: TInput) => TOutput; // <-- this is the one
```

Looking at the existing code, this would actually require modifying `setQueryData`, `functionalUpdate`, and probably others too, but having access to the full `queryKey` in the updater function would be very beneficial (either as the second argument, or as a property of the second argument, so adding more items than `queryKey` wouldn't constitute a breaking change in the future).

@TkDodo would this be something you'd consider? For the moment, we'd have to resort to the query key iteration you wrote in an earlier comment.

**Reply by TkDodo** [reply]
I'm not sure I understand the problem. Given that `filteredItems` and `moreFilteredItems` have the same structure, what I would do is try to find the item to update in all matching entries and update that one. If the item I'm updating exists in more than one entry, it would get updated in all of them. Isn't that what we want?

**Reply by bartvanandel** [reply]
Well, no. `state` has changed in the item, so it should be *removed* from `filteredItems`, and added to `moreFilteredItems`. But since the updater function has no knowledge of the `queryKey`, it cannot know this.

**Reply by TkDodo** [reply]
I see now. Hm, it's not something I would personally do, because the complexity is quite high. If you know the key, you would have to derive from the key that you have to either insert / update / delete an item. If you can do that, cool. But at this point, I'm duplicating so much server logic that I'd rather just invalidate.

But, I'm not against that change. Should the `updater` just take the whole `query` as a second argument?

**Reply by bartvanandel** [reply]
I think that makes sense indeed, as it provides the most flexibility. Thanks!

**Reply by TkDodo** [reply]
would you like to contribute that?

**Reply by bartvanandel** [reply]
Sure, I can give it a go. Let's see if I can find some time tonight or tomorrow.

**Reply by basslagter** [reply]
Hi, 

I stumbled upon this thread in my search of (re)setting the cache for a specific key to undefined.
removeQueries, deleteQueries all do not seem te have the desired effect. I found that this piece of code does the trick:

```
   queryClient
      .getQueryCache()
      .findAll({ queryKey })
      .forEach(query => query.setData(undefined));
```

But that seems a bit complex. Then I saw this 'setQueriesData' but this:

`queryClient.setQueriesData({ queryKey }, () => undefined);`

has not the same effect as the code above, which seems weird.

To clearify; in my situation I have once fetched some data for a specific query key. At some point I get an event about that data being no longer relevant thus I want to clear the cache for that key (or make it undefined) so my components will all rerender and show their "undefined" cases.

**Reply by YassienW** [reply]
@basslagter According to the [docs](https://tanstack.com/query/v5/docs/react/reference/QueryClient#queryclientsetquerydata), `setQueryData` doesn't update the value if undefined is returned. `setQueriesData` calls this function under the hood, which is causing your issue. [resetQueries](https://tanstack.com/query/latest/docs/react/reference/QueryClient#queryclientresetqueries) is probably what you're looking for.

**Reply by TkDodo** [reply]
> At some point I get an event about that data being no longer relevant thus I want to clear the cache for that key (or make it undefined) so my components will all rerender and show their "undefined" cases.

don't set the query to `undefined`, as `undefined` is not a valid value for a query in success state. You can set it t `null` if you want.

what `resetQueries` will do is reset the query to it's initial state, which is `pending`, in which case it will go and fetch latest data. Not sure if that's what you want

**Reply by basslagter** [reply]
Ah, well....that makes sense although I did not quite get this from the docs tbh. But, setting it to `null` works perfectly, thanks!
`queryClient.setQueriesData({ queryKey }, () => null)`

**Reply by TkDodo** [reply]
It's documented here: https://tanstack.com/query/v5/docs/react/reference/QueryClient#queryclientsetquerydata

**Reply by danielgran** [reply]
but.... how?

**Reply by basslagter** [reply]
I have an additional question. It turns out that when I do `queryClient.setQueriesData({ queryKey }, () => null)` the data for that key is indeed reset. But...when I now do a 'fetch' on the same key, it is not calling my API anymore. Probably because the key is still in the cache (with null as value). Can I remove the entry entirely?

**Reply by TkDodo** [reply]
> Can I remove the entry entirely?

`queryClient.removeQueries`

**Reply by IgorKurkov** [reply]
As @bartvanandel said the `setQueriesData` won't work with NOT exact key. I used `.getQueryCache().findAll....` approach as a viable one (as a workaround), but idk what to do with setQueriesData if it presented but doesn't work with {exact: false} keys? Sorry for boring question, but do we have any feature/bug request for it?

**Reply by YassienW** [reply]
@IgorKurkov [Query filters](https://tanstack.com/query/latest/docs/framework/react/guides/filters#query-filters) accept an arbitrary filter function `predicate`

**Reply by TkDodo** [reply]
> the setQueriesData won't work with NOT exact key. I used .getQueryCache().findAll.... approach as a viable one (as a workaround)

`setQueriesData` is literally implemented with `queryCache.findAll`:

https://github.com/TanStack/query/blob/febbde676d0888d8dd4868177868a3316686cb58/packages/query-core/src/queryClient.ts#L206-L219
